File: core/py/src/archon_core/_conformance.py
"""Conformance CLI (python) — a dev/CI artifact, not part of the library surface.

Implements the `conformance v1` protocol: `conformance <family>` reads the whole
vectors/identity.json on stdin, selects its family's cases, RECOMPUTES each result from the
case INPUTS (ignoring the expected value the oracle carries), and writes one NDJSON line per
case to stdout in input order. The harness drives this and the other cores as black boxes and
asserts each line against the oracle.

    pubkey_from_seed : in {name, seed}                  out {"name","pubkey":"<64-hex>"}
    key_encode       : in {name, pubkey}                out {"name","text":"<key text>"}
    keycodec         : in {name, kind, key?|pem?}       out {"name","result":{"ok":…}|{"error":true}}
    signature_verify : in {name, pubkey, message, sig}  out {"name","valid":<bool>}
    hex_decode       : in {name, kind, hex}             out {"name","result":{"ok":"<hex>"}|{"error":true}}
    domain_sign      : in {name, seed, domain, message} out {"name","result":{"ok":"<128-hex>"}|{"error":true}}
    domain_verify    : in {name, pubkey, domain, message, sig} out {"name","valid":<bool>}

Recomputing rather than echoing is the whole point: a CLI that read `result` from the case
would agree with the oracle by construction and prove nothing.
"""

import json
import sys

from . import crypto, hexbytes, keycodec, keytext


def _result_of(value, err):
    """The oracle's two-shape result: {"ok": …} or {"error": true}."""
    return {"error": True} if err is not None else {"ok": value}


def _try(fn):
    try:
        return fn(), None
    except Exception as exc:  # noqa: BLE001 - every failure is one refusal to the oracle
        return None, exc


def _unhex(text: str) -> bytes:
    # Case inputs that are *already* known-good hex; a bad one is a broken vector file,
    # not a case the family is testing.
    return bytes.fromhex(text)


def main(argv=None) -> int:
    argv = sys.argv[1:] if argv is None else argv
    if len(argv) != 1:
        print("usage: conformance <family>", file=sys.stderr)
        return 2
    family = argv[0]

    doc = json.load(sys.stdin)
    cases = doc.get(family)
    if cases is None:
        print(f"conformance: no such family: {family}", file=sys.stderr)
        return 1

    out_lines = []
    for case in cases:
        # The oracle carries a leading note object in some families; it has no name.
        if "name" not in case:
            continue
        out = {"name": case["name"]}

        if family == "pubkey_from_seed":
            pub = crypto.public_key_from_seed(_unhex(case["seed"]))
            out["pubkey"] = hexbytes.to_hex(pub)

        elif family == "key_encode":
            out["text"] = keytext.encode_key(_unhex(case["pubkey"]))

        elif family == "keycodec":
            kind = case["kind"]
            if kind == "encode_pkcs8":
                v, e = _try(lambda: keycodec.seed_to_pkcs8_pem(_unhex(case["key"])).decode("ascii"))
            elif kind == "encode_spki":
                v, e = _try(lambda: keycodec.pubkey_to_spki_pem(_unhex(case["key"])).decode("ascii"))
            elif kind == "decode_pkcs8":
                v, e = _try(lambda: hexbytes.to_hex(keycodec.pkcs8_pem_to_seed(case["pem"].encode("utf-8"))))
            elif kind == "decode_spki":
                v, e = _try(lambda: hexbytes.to_hex(keycodec.spki_pem_to_pubkey(case["pem"].encode("utf-8"))))
            else:
                raise SystemExit(f"unknown keycodec kind: {kind}")
            out["result"] = _result_of(v, e)

        elif family == "signature_verify":
            out["valid"] = crypto.verify(
                _unhex(case["pubkey"]), _unhex(case["message"]), _unhex(case["sig"])
            )

        elif family == "hex_decode":
            kind = case["kind"]
            fn = {
                "seed": hexbytes.seed_from_hex,
                "pubkey": hexbytes.pubkey_from_hex,
                "signature": hexbytes.signature_from_hex,
            }.get(kind)
            if fn is None:
                raise SystemExit(f"unknown hex_decode kind: {kind}")
            v, e = _try(lambda: hexbytes.to_hex(fn(case["hex"])))
            out["result"] = _result_of(v, e)

        elif family == "domain_sign":
            v, e = _try(
                lambda: hexbytes.to_hex(
                    crypto.sign_in_domain(
                        _unhex(case["seed"]), case["domain"], _unhex(case["message"])
                    )
                )
            )
            out["result"] = _result_of(v, e)

        elif family == "domain_verify":
            out["valid"] = crypto.verify_in_domain(
                _unhex(case["pubkey"]),
                case["domain"],
                _unhex(case["message"]),
                _unhex(case["sig"]),
            )

        else:
            print(f"conformance: unhandled family: {family}", file=sys.stderr)
            return 1

        out_lines.append(json.dumps(out, separators=(",", ":"), sort_keys=True))

    sys.stdout.write("\n".join(out_lines) + ("\n" if out_lines else ""))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: core/rs/src/bin/conformance.rs
//! Conformance CLI (rs) — a dev/CI artifact, NOT part of the published lib API. It is
//! built only under the `conformance-cli` feature, so the default build keeps archon's
//! austerity: `ed25519-dalek` and nothing else.
//!
//! Implements the `conformance v1` protocol archon inherits from thesmos ADR 0006:
//! `conformance <family>` reads the whole vectors/identity.json on stdin, selects its
//! family's cases, recomputes each result from the case INPUTS (ignoring the expected
//! value the oracle carries), and writes one NDJSON line per case to stdout in input
//! order. The language-agnostic harness (`conformance/harness.mjs`) drives this and the
//! go/ts CLIs as black boxes and asserts each line against the oracle.
//!
//!   pubkey_from_seed : in `{name, seed}`                out `{"name","pubkey":"<64-hex>"}`
//!   key_encode       : in `{name, pubkey}`              out `{"name","text":"<key text>"}`
//!   keycodec         : in `{name, kind, key?|pem?}`     out `{"name","result":{"ok":"<PEM|hex>"}|{"error":true}}`
//!   signature_verify : in `{name, pubkey, message, sig}` out `{"name","valid":<bool>}`
//!   hex_decode       : in `{name, kind, hex}`            out `{"name","result":{"ok":"<hex>"}|{"error":true}}`
//!   domain_sign      : in `{name, seed, domain, message}` out `{"name","result":{"ok":"<128-hex>"}|{"error":true}}`
//!   domain_verify    : in `{name, pubkey, domain, message, sig}` out `{"name","valid":<bool>}`

use archon_core::{crypto, hexbytes, keycodec, keytext};
use serde_json::{json, Value};
use std::io::Read;

fn hex_decode(s: &str) -> Vec<u8> {
    (0..s.len())
        .step_by(2)
        .map(|i| u8::from_str_radix(&s[i..i + 2], 16).expect("case input is not valid hex"))
        .collect()
}

fn hex_encode(b: &[u8]) -> String {
    b.iter().map(|x| format!("{x:02x}")).collect()
}

/// Every codec conversion returns Result<_, String>; the oracle only distinguishes
/// success-with-a-value from failure, never the message, so errors collapse to
/// `{"error":true}`. That is deliberate: the reject *reason* is a core's own diagnostic,
/// the reject *decision* is what must agree byte-for-byte across three languages.
fn result_json(r: Result<String, String>) -> Value {
    match r {
        Ok(v) => json!({ "ok": v }),
        Err(_) => json!({ "error": true }),
    }
}

fn main() {
    let family = std::env::args().nth(1).unwrap_or_else(|| {
        eprintln!("usage: conformance <family>");
        std::process::exit(2);
    });

    let mut input = String::new();
    std::io::stdin()
        .read_to_string(&mut input)
        .expect("read stdin");
    let doc: Value = serde_json::from_str(&input).expect("stdin is not valid JSON");
    let cases = doc
        .get(&family)
        .and_then(Value::as_array)
        .unwrap_or_else(|| panic!("unknown family: {family}"));

    for c in cases {
        let name = c["name"].as_str().unwrap_or("").to_string();
        let line = match family.as_str() {
            "pubkey_from_seed" => {
                let seed: [u8; 32] = hex_decode(c["seed"].as_str().unwrap())
                    .try_into()
                    .expect("seed is not 32 bytes");
                json!({ "name": name, "pubkey": hex_encode(&crypto::public_key_from_seed(&seed)) })
            }
            "key_encode" => {
                let pubkey = hex_decode(c["pubkey"].as_str().unwrap());
                json!({ "name": name, "text": keytext::encode_key(&pubkey) })
            }
            "keycodec" => {
                let kind = c["kind"].as_str().unwrap();
                let r = match kind {
                    "encode_pkcs8" => {
                        keycodec::seed_to_pkcs8_pem(&hex_decode(c["key"].as_str().unwrap()))
                    }
                    "encode_spki" => {
                        keycodec::pubkey_to_spki_pem(&hex_decode(c["key"].as_str().unwrap()))
                    }
                    "decode_pkcs8" => keycodec::pkcs8_pem_to_seed(c["pem"].as_str().unwrap())
                        .map(|k| hex_encode(&k)),
                    "decode_spki" => keycodec::spki_pem_to_pubkey(c["pem"].as_str().unwrap())
                        .map(|k| hex_encode(&k)),
                    other => panic!("unknown keycodec kind: {other}"),
                };
                json!({ "name": name, "result": result_json(r) })
            }
            "signature_verify" => {
                let pubkey = hex_decode(c["pubkey"].as_str().unwrap());
                let message = hex_decode(c["message"].as_str().unwrap());
                let sig = hex_decode(c["sig"].as_str().unwrap());
                json!({ "name": name, "valid": crypto::verify(&pubkey, &message, &sig) })
            }
            "hex_decode" => {
                let text = c["hex"].as_str().unwrap();
                let r = match c["kind"].as_str().unwrap() {
                    "seed" => hexbytes::seed_from_hex(text).map(|b| hexbytes::to_hex(&b)),
                    "pubkey" => hexbytes::pubkey_from_hex(text).map(|b| hexbytes::to_hex(&b)),
                    "signature" => hexbytes::signature_from_hex(text).map(|b| hexbytes::to_hex(&b)),
                    other => panic!("unknown hex_decode kind: {other}"),
                };
                json!({ "name": name, "result": result_json(r) })
            }
            "domain_sign" => {
                let seed: [u8; 32] = hex_decode(c["seed"].as_str().unwrap())
                    .try_into()
                    .expect("seed is not 32 bytes");
                let message = hex_decode(c["message"].as_str().unwrap());
                let r = crypto::sign_in_domain(&seed, c["domain"].as_str().unwrap(), &message)
                    .map(|s| hex_encode(&s));
                json!({ "name": name, "result": result_json(r) })
            }
            "domain_verify" => {
                let pubkey = hex_decode(c["pubkey"].as_str().unwrap());
                let message = hex_decode(c["message"].as_str().unwrap());
                let sig = hex_decode(c["sig"].as_str().unwrap());
                let domain = c["domain"].as_str().unwrap();
                json!({ "name": name, "valid": crypto::verify_in_domain(&pubkey, domain, &message, &sig) })
            }
            other => panic!("unknown family: {other}"),
        };
        println!("{line}");
    }
}
